File: apps/desktop/src/lib/index-activity.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import { CURRENT_PROTOCOL_VERSION, type InvokeCommand } from './health';

export interface QueryIndexActivityRequestV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
}

export type IndexActivityStateV1 =
  'idle' | 'queued' | 'running' | 'cancelling' | 'succeeded' | 'failed' | 'cancelled';
export type IndexPhaseV1 = 'discover' | 'hash' | 'parse' | 'link' | 'rank' | 'publish';

export interface IndexActivityV1 {
  completedPhases: number;
  phase: IndexPhaseV1 | null;
  state: IndexActivityStateV1;
  totalPhases: 6;
}

export type IndexActivityResultV1 =
  { status: 'noProject' } | { activity: IndexActivityV1; status: 'active' };

export interface IndexActivityResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: IndexActivityResultV1;
}

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function queryIndexActivity(
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<IndexActivityResponseV1> {
  const request: QueryIndexActivityRequestV1 = { protocolVersion: CURRENT_PROTOCOL_VERSION };
  const payload = await invokeCommand('query_index_activity', { request });
  return parseIndexActivityResponseV1(payload);
}

export function parseIndexActivityResponseV1(payload: unknown): IndexActivityResponseV1 {
  if (
    !isRecord(payload) ||
    !hasExactKeys(payload, ['protocolVersion', 'result']) ||
    payload.protocolVersion !== CURRENT_PROTOCOL_VERSION
  ) {
    throw new Error('Index activity response does not match the V1 schema.');
  }

  return {
    protocolVersion: payload.protocolVersion,
    result: parseResult(payload.result),
  };
}

function parseResult(value: unknown): IndexActivityResultV1 {
  if (!isRecord(value) || typeof value.status !== 'string') {
    throw new Error('Index activity response contains an invalid result.');
  }
  if (value.status === 'noProject' && hasExactKeys(value, ['status'])) {
    return { status: 'noProject' };
  }
  if (value.status === 'active' && hasExactKeys(value, ['activity', 'status'])) {
    return { activity: parseActivity(value.activity), status: 'active' };
  }
  throw new Error('Index activity response contains an invalid result.');
}

function parseActivity(value: unknown): IndexActivityV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ['completedPhases', 'phase', 'state', 'totalPhases']) ||
    !isActivityState(value.state) ||
    !Number.isInteger(value.completedPhases) ||
    typeof value.completedPhases !== 'number' ||
    value.completedPhases < 0 ||
    value.completedPhases > 6 ||
    value.totalPhases !== 6 ||
    (value.phase !== null && !isIndexPhase(value.phase))
  ) {
    throw new Error('Index activity response contains invalid progress.');
  }

  if (value.state === 'idle') {
    if (value.phase !== null || value.completedPhases !== 0) {
      throw new Error('Idle index activity contains contradictory progress.');
    }
  } else {
    if (value.phase === null || !phaseMatchesProgress(value.phase, value.completedPhases)) {
      throw new Error('Index activity phase does not match its progress.');
    }
    if (value.state === 'queued' && (value.phase !== 'discover' || value.completedPhases !== 0)) {
      throw new Error('Queued index activity must begin with discovery.');
    }
    if (value.state === 'succeeded' && value.completedPhases !== 6) {
      throw new Error('Succeeded index activity must be complete.');
    }
  }

  return {
    completedPhases: value.completedPhases,
    phase: value.phase,
    state: value.state,
    totalPhases: value.totalPhases,
  };
}

function phaseMatchesProgress(phase: IndexPhaseV1, completed: number): boolean {
  const expected: Record<IndexPhaseV1, number> = {
    discover: 0,
    hash: 1,
    parse: 2,
    link: 3,
    rank: 4,
    publish: 5,
  };
  return completed === expected[phase] || (phase === 'publish' && completed === 6);
}

function isActivityState(value: unknown): value is IndexActivityStateV1 {
  return (
    value === 'idle' ||
    value === 'queued' ||
    value === 'running' ||
    value === 'cancelling' ||
    value === 'succeeded' ||
    value === 'failed' ||
    value === 'cancelled'
  );
}

function isIndexPhase(value: unknown): value is IndexPhaseV1 {
  return (
    value === 'discover' ||
    value === 'hash' ||
    value === 'parse' ||
    value === 'link' ||
    value === 'rank' ||
    value === 'publish'
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const keys = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    keys.length === sortedExpected.length &&
    keys.every((key, index) => key === sortedExpected[index])
  );
}

File: crates/a3-protocol/src/index_activity.rs
use crate::ProtocolVersion;
use serde::{Deserialize, Serialize};

/// Strict input payload for the lightweight V1 Fast-Index activity query.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryIndexActivityRequestV1 {
    protocol_version: ProtocolVersion,
}

impl QueryIndexActivityRequestV1 {
    /// Creates a request for a specific protocol version.
    #[must_use]
    pub const fn new(protocol_version: ProtocolVersion) -> Self {
        Self { protocol_version }
    }

    /// Creates a request for the protocol version emitted by this build.
    #[must_use]
    pub const fn current() -> Self {
        Self::new(ProtocolVersion::CURRENT)
    }

    /// Returns the requested protocol version.
    #[must_use]
    pub const fn protocol_version(self) -> ProtocolVersion {
        self.protocol_version
    }
}

/// Lightweight response that never reads repository files or reconstructs an index.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct IndexActivityResponseV1 {
    protocol_version: ProtocolVersion,
    result: IndexActivityResultV1,
}

impl IndexActivityResponseV1 {
    /// Creates the response used before any project has been selected.
    #[must_use]
    pub const fn no_project() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: IndexActivityResultV1::NoProject,
        }
    }

    /// Creates a response from the Core-owned activity snapshot.
    #[must_use]
    pub const fn active(activity: IndexActivityV1) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: IndexActivityResultV1::Active { activity },
        }
    }

    /// Returns the mutually exclusive project/activity result.
    #[must_use]
    pub const fn result(&self) -> &IndexActivityResultV1 {
        &self.result
    }
}

/// Whether a Core-owned project exists for the activity query.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum IndexActivityResultV1 {
    /// No project is active in this desktop process.
    NoProject,
    /// The bounded scheduler projection for the active project.
    Active {
        /// Current Fast-Index lifecycle and phase.
        activity: IndexActivityV1,
    },
}

/// Bounded projection of one owned Fast-Index job.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct IndexActivityV1 {
    state: IndexActivityStateV1,
    phase: Option<IndexPhaseV1>,
    completed_phases: u64,
    total_phases: u64,
}

impl IndexActivityV1 {
    /// Creates a projection from manager-validated lifecycle values.
    #[must_use]
    pub const fn new(
        state: IndexActivityStateV1,
        phase: Option<IndexPhaseV1>,
        completed_phases: u64,
        total_phases: u64,
    ) -> Self {
        Self {
            state,
            phase,
            completed_phases,
            total_phases,
        }
    }

    /// Returns the scheduler-owned lifecycle state.
    #[must_use]
    pub const fn state(self) -> IndexActivityStateV1 {
        self.state
    }

    /// Returns the deterministic phase, if a run has started.
    #[must_use]
    pub const fn phase(self) -> Option<IndexPhaseV1> {
        self.phase
    }

    /// Returns completed phase boundaries.
    #[must_use]
    pub const fn completed_phases(self) -> u64 {
        self.completed_phases
    }

    /// Returns the fixed V1 phase count.
    #[must_use]
    pub const fn total_phases(self) -> u64 {
        self.total_phases
    }
}

/// Lifecycle of the current or most recently completed Fast-Index job.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum IndexActivityStateV1 {
    /// No Fast-Index job has been submitted for the active project.
    Idle,
    /// The bounded scheduler accepted the job.
    Queued,
    /// An owned worker is executing the job.
    Running,
    /// Cooperative cancellation was requested.
    Cancelling,
    /// The job completed and its publication is visible.
    Succeeded,
    /// The job failed without replacing the previous publication.
    Failed,
    /// The job stopped cooperatively without replacing the previous publication.
    Cancelled,
}

/// Fixed ADR-0006 Fast-Index phase names.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum IndexPhaseV1 {
    /// Discover repository candidates.
    Discover,
    /// Hash exact contents.
    Hash,
    /// Parse supported files.
    Parse,
    /// Link graph relationships.
    Link,
    /// Rank symbols and form modules.
    Rank,
    /// Atomically publish the complete index.
    Publish,
}

#[cfg(test)]
mod tests {
    use super::{IndexActivityResponseV1, IndexActivityStateV1, IndexActivityV1, IndexPhaseV1};
    use serde_json::json;

    #[test]
    fn active_activity_has_a_strict_bounded_shape() -> Result<(), serde_json::Error> {
        let response = IndexActivityResponseV1::active(IndexActivityV1::new(
            IndexActivityStateV1::Running,
            Some(IndexPhaseV1::Link),
            3,
            6,
        ));

        assert_eq!(
            serde_json::to_value(response)?,
            json!({
                "protocolVersion": 1,
                "result": {
                    "status": "active",
                    "activity": {
                        "state": "running",
                        "phase": "link",
                        "completedPhases": 3,
                        "totalPhases": 6
                    }
                }
            })
        );
        Ok(())
    }
}
